cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -1388,5 +1388,9 @@ private class IntegerOrPointerConstantInstruction extends ConstantInstruction {
 
 /** The int value of integer constant expression. */
 private int int_value(Instruction i) {
-  result = valueNumber(i).getAnInstruction().(IntegerOrPointerConstantInstruction).getValue().toInt()
+  result =
+    valueNumber(i).getAnInstruction().(IntegerOrPointerConstantInstruction).getValue().toInt()
+  or
+  // handle conversions
+  result = int_value(valueNumber(i).getAnInstruction().(ConvertInstruction).getUnary())
```

Comment:
Watch out with materializing all `Instruction`s when all we really care about is the value number. If we instead change the `i` parameter to be a `ValueNumber` instead this predicate will be a lot smaller.

Note that this will require changing all the callers so that they convert the argument to a value number. You can define an (inline) helper predicate to do that conversion for you. This will keep all callers as-is, and still avoid unnecessary materialization

Reply:
good point, but what about inlining int_value instead?

Reply:
You can't inline it since it's recursive, unfortunately.

Reply:
Ah true true. The inline helper function I think is the winner then, which is basically the best of both worlds.